### PR TITLE
[action] setup_ci – always setup keychain when running on a CI

### PR DESCRIPTION
**Pull request by svenmuennich**
### Checklist
- [x] I've run `bundle exec rspec` from the root directory to see all new and existing tests pass
- [x] I've followed the _fastlane_ code style and run `bundle exec rubocop -a` to ensure the code style is valid
- [x] I've read the [Contribution Guidelines](https://github.com/fastlane/fastlane/blob/master/CONTRIBUTING.md)
- [x] I've updated the documentation if necessary.

### Motivation and Context
Currently `setup_ci` must be passed a valid `:provider` (either `travis` or `circleci`) that controls which setup is performed for each of them. However, the only difference between the providers is that passing `circleci` also configures the output paths, while the temporary keychain is created for both of them.

As pointed out in #15445, running `setup_ci` in GitHub Actions without passing one of the valid providers currently results in a noop, even though the environment is recognized as being a CI. However, one would expect that the action also works on other CI providers and e.g. creates a temporary keychain.

### Description
Based on the @tlvenn's proposal put forward in #15445 I changed `setup_ci` to always create a temporary keychain, but only after performing any extra tasks necessary for a specific CI provider.

Furthermore the CI provider is now detected automatically, if no `:provider` is passed. I left the validation of a passed `:provider` in place to ensure backwards compatibility, even though passing `travis` does  not change the behavior of the action compared to not passing any `:provider`.

To not having to duplicate parts of the CI detection I added a new method `Helper.is_circle_ci`, that is used both in `Helper.ci?` and `setup_ci`.

### CODE REVIEW

**Comment by svenmuennich** [comment]
@googlebot I signed it!

**Comment by janpio** [comment]
> @googlebot I signed it!

That didn't work it seems. Usually this is because your GitHub account uses a different email address or the commits were made using a different email address.

**Comment by svenmuennich** [comment]
> That didn't work it seems. Usually this is because your GitHub account uses a different email address or the commits were made using a different email address.

I think this is due to the company CLA we signed last week. The verification process has not yet completed.

File: fastlane/lib/fastlane/actions/setup_ci.rb [hunk 1]
```diff
@@ -7,19 +7,22 @@ def self.run(params)
...
         end
+
+        setup_keychain
       end
 
       def self.should_run?(params)
         Helper.ci? || params[:force]
       end
 
+      def self.detect_provider(params)
```
Comment (tlvenn) [inline fastlane/lib/fastlane/actions/setup_ci.rb]: I feel this is beyond the scope of the PR and I feel it's weird that circleci receive a special treatment there. Either we can flirt the CI and then we should do it or we should keep the current behavior.

File: fastlane/lib/fastlane/actions/setup_ci.rb [hunk 2]
```diff
@@ -7,19 +7,22 @@ def self.run(params)
...
-        case params[:provider]
-        when 'travis'
-          setup_keychain
+        case detect_provider(params)
         when 'circleci'
-          setup_keychain
           setup_output_paths
         end
+
+        setup_keychain
```
Comment (tlvenn) [inline fastlane/lib/fastlane/actions/setup_ci.rb]: `setup_keychain` was always called first, this is now called last which seems wrong to me even if probably it does not matter in the case of the additional `setup_output_paths` for Travis. However i believe there is a test which expect a given order and this is breaking.

File: fastlane/lib/fastlane/actions/setup_ci.rb [hunk 3]
```diff
@@ -87,18 +90,20 @@ def self.available_options
                                        default_value: false),
           FastlaneCore::ConfigItem.new(key: :provider,
                                        env_name: "FL_SETUP_CI_PROVIDER",
-                                       description: "CI provider",
+                                       description: "CI provider. If none is set, the provider is detected automatically",
```
Comment (tlvenn) [inline fastlane/lib/fastlane/actions/setup_ci.rb]: This is misleading as only circleci would be detected and imho as discussed above, this is outside of the scope of what this PR should do.

File: fastlane/lib/fastlane/actions/setup_ci.rb [hunk 4]
```diff
@@ -108,7 +113,7 @@ def self.is_supported?(platform)
       def self.example_code
         [
           'setup_ci(
-            provider: "travis"
+            provider: "circleci"
```
Comment (tlvenn) [inline fastlane/lib/fastlane/actions/setup_ci.rb]: Any reason to change the example ?

File: fastlane/lib/fastlane/actions/setup_ci.rb (hunk 2 above)
Comment (tlvenn) [reply fastlane/lib/fastlane/actions/setup_ci.rb]: Actually as you commented Message is asserted in reverse order so the test would pass but still I wonder if conceptually the action should not do first what it is supposed to do unless there is a very good reason to do something else first.

**Review by tlvenn — COMMENTED** [review]
Thanks for the PR, I have mixed feeling with this one, I feel you are blending too much circleci stuff that could / should ? be in a separate PR.

File: fastlane/lib/fastlane/actions/setup_ci.rb (hunk 1 above)
Comment (svenmuennich) [reply fastlane/lib/fastlane/actions/setup_ci.rb]: Since the gist of this PR is that running `setup_ci` always installs the keychain, passing a `provider` param should be optional. That said, it might appear like _CircleCI_ receives special treatment but that is just because it is the only CI provider for which other setup step need to be performed. I didn't see the point in detecting every CI provider but then not acting on the result.

File: fastlane/lib/fastlane/actions/setup_ci.rb (hunk 2 above)
Comment (svenmuennich) [reply fastlane/lib/fastlane/actions/setup_ci.rb]: I actually just followed the suggestion made in your issue #15445:
> If a CI is detected, setup_keychain should be called in all cases **and only after** [...] **more tasks could be executed** with a case/when statement.

So I really don't mind moving this before the `case/when` statement.

File: fastlane/lib/fastlane/actions/setup_ci.rb (hunk 3 above)
Comment (svenmuennich) [reply fastlane/lib/fastlane/actions/setup_ci.rb]: Please see my explanation above.

File: fastlane/lib/fastlane/actions/setup_ci.rb (hunk 4 above)
Comment (svenmuennich) [reply fastlane/lib/fastlane/actions/setup_ci.rb]: I changed this to to prevent false assumptions that passing `travisci` changed the behaviour of this action, which it doesn't.

File: fastlane/lib/fastlane/actions/setup_ci.rb (hunk 2 above)
Comment (tlvenn) [reply fastlane/lib/fastlane/actions/setup_ci.rb]: I meant to say only after calling `setup_keychain`, other actions should be called with...
Sorry for the confusion, I see how this could be interpreted the other way around.

**Comment by svenmuennich** [comment]
@googlebot I signed it!

**Review by joshdholtz — APPROVED** [review]
Thanks for making this change! I think these changes make a lot more sense. Appreciate the contribution ❤️
